In [1]:
#필요한 모듈 호출
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.svm import SVC
from sklearn.model_selection import train_test_split
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler

#IRIS 데이터 로드
iris =  sns.load_dataset('iris') 
X= iris.iloc[:,:4] #학습할데이터
y = iris.iloc[:,-1] #타겟
print(y)

0         setosa
1         setosa
2         setosa
3         setosa
4         setosa
         ...    
145    virginica
146    virginica
147    virginica
148    virginica
149    virginica
Name: species, Length: 150, dtype: object


In [28]:
X #array로 저장됨

array([[-9.00681170e-01,  1.01900435e+00, -1.34022653e+00,
        -1.31544430e+00],
       [-1.14301691e+00, -1.31979479e-01, -1.34022653e+00,
        -1.31544430e+00],
       [-1.38535265e+00,  3.28414053e-01, -1.39706395e+00,
        -1.31544430e+00],
       [-1.50652052e+00,  9.82172869e-02, -1.28338910e+00,
        -1.31544430e+00],
       [-1.02184904e+00,  1.24920112e+00, -1.34022653e+00,
        -1.31544430e+00],
       [-5.37177559e-01,  1.93979142e+00, -1.16971425e+00,
        -1.05217993e+00],
       [-1.50652052e+00,  7.88807586e-01, -1.34022653e+00,
        -1.18381211e+00],
       [-1.02184904e+00,  7.88807586e-01, -1.28338910e+00,
        -1.31544430e+00],
       [-1.74885626e+00, -3.62176246e-01, -1.34022653e+00,
        -1.31544430e+00],
       [-1.14301691e+00,  9.82172869e-02, -1.28338910e+00,
        -1.44707648e+00],
       [-5.37177559e-01,  1.47939788e+00, -1.28338910e+00,
        -1.31544430e+00],
       [-1.26418478e+00,  7.88807586e-01, -1.22655167e+00,
      

In [17]:
#타겟변수 종류 살피기
y.unique()

array(['setosa', 'versicolor', 'virginica'], dtype=object)

3종류

In [2]:
#스케일링 해주기
scal = StandardScaler() 
X = scal.fit_transform(X)

In [4]:
print(y)

0         setosa
1         setosa
2         setosa
3         setosa
4         setosa
         ...    
145    virginica
146    virginica
147    virginica
148    virginica
149    virginica
Name: species, Length: 150, dtype: object


In [5]:
# One Versus Rest 구현을 위해 class가 0인것과 아닌것을 구분하기위한 machine class가 1인것과 아닌 것을 구분하기 위한 머신 등을 미리만들어줌
svm_1 = SVC(kernel ='rbf', C = 5, gamma = 5) #setosa
svm_2 = SVC(kernel ='rbf', C = 5, gamma = 5) #versicolor
svm_3 = SVC(kernel ='rbf', C = 5, gamma = 5) #virginica

## from sklearn.model_selection import train_test_split #테스트/트레인 데이터로 분리
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=48)

In [6]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=48)

In [7]:
y_train = pd.get_dummies(y_train) #one hot encoding을 먼저해보자

In [16]:
svm_1.fit(X_train,y_train.iloc[:,0]) # 데이터 클레스가 0인 애인지 아닌지 구분해주는 머신
svm_2.fit(X_train,y_train.iloc[:,1]) # 데이터 클레스가 1인 애인지 아닌지 구분해주는 머신
svm_3.fit(X_train,y_train.iloc[:,2]) # 데이터 클레스가 2인 애인지 아닌지 구분해주는 머신
print(svm_1.predict(X_test)) #데이터 클래스가 0인지 아닌지 구분해주는 애를 통해서 테스트데이터의 클래스가 0인지 0이 아닌인지 예측해보자

print(svm_1.decision_function(X_test)) #decision_function hyperplane과의 거리를 구하는 방법 
print(svm_2.decision_function(X_test)) #svm_2와 svm_3에도 적용해서 한꺼번에 프린트해보기
print(svm_3.decision_function(X_test))

[0 0 0 0 0 0 1 0 1 0 0 1 1 0 0 0 1 0 0 0 1 0 0 0 0 1 1 0 0 0]
[-1.12470845 -0.86326953 -0.65281099 -0.50248821 -0.76284323 -0.87465573
  1.07709345 -0.99281647  0.47441336 -0.99842743 -0.83989348  0.15633457
  0.32871788 -0.97965464 -0.72385083 -0.92638376  1.28322481 -0.56240455
 -0.72719892 -0.99509775  0.43166724 -0.96451557 -0.82991366 -1.03020581
 -0.75166835  1.13461335  0.39943974 -1.04194106 -0.93376548 -1.06133798]
[ 1.36054602  0.62859488  0.0580828  -0.36987093 -0.25686625 -0.79229874
 -1.03195247 -0.87132049 -0.77957234  0.99327762 -0.73309581 -0.64607768
 -0.71843897 -0.80882458  0.47748663  0.69953462 -1.11879121 -0.1870504
 -0.24458236 -0.99337352 -0.76149141 -0.96812018  0.11310267  1.06892719
 -0.56749692 -1.05639624 -0.74817515 -1.0537858  -0.89721836  1.17126024]
[-1.26277651 -0.77964936 -0.40539105 -0.13070947  0.02643314  0.57365649
 -1.04443118  0.88027844 -0.69605517 -0.99074328  0.52442024 -0.51171892
 -0.61157919  0.78148257 -0.75517707 -0.76905925 -1.16380018 

In [18]:
#테스트데이터 개수 구하기
len(svm_1.decision_function(X_test))

30

In [25]:
## Case 1 : One vs Rest SVM을 이부분에 구현해주세요 위 결과들을 이용해서 multi class SVM을 직접 구현해주세요! 하드코딩이 하시기 편할겁니다.

# 결정함수값이 양수, 음수인지에 따라 타겟변수 3종류 중 하나를 예측하는 함수
group = []
for i in range(30):
    if svm_1.decision_function(X_test)[i] >0 and svm_2.decision_function(X_test)[i] <0 and svm_3.decision_function(X_test)[i] <0:
        group.append('setosa') # svm_1(+),svm_2(-),svm_3(-)이면 setosa
    elif svm_1.decision_function(X_test)[i] <0 and svm_2.decision_function(X_test)[i] >0 and svm_3.decision_function(X_test)[i] <0:
        group.append('versicolor') # svm_1(-),svm_2(+),svm_3(-)이면 versicolor
    elif svm_1.decision_function(X_test)[i] <0 and svm_2.decision_function(X_test)[i] <0 and svm_3.decision_function(X_test)[i] >0:
        group.append('virginica') # svm_1(-),svm_2(-),svm_3(+)이면 virginica
    else: # 나머지 경우는 0으로 표시
        group.append(0)

group

['versicolor',
 'versicolor',
 'versicolor',
 0,
 'virginica',
 'virginica',
 'setosa',
 'virginica',
 'setosa',
 'versicolor',
 'virginica',
 'setosa',
 'setosa',
 'virginica',
 'versicolor',
 'versicolor',
 'setosa',
 0,
 0,
 'virginica',
 'setosa',
 'virginica',
 'versicolor',
 'versicolor',
 'virginica',
 'setosa',
 'setosa',
 'virginica',
 'virginica',
 'versicolor']

분류 안되는 4번째,18번째,19번째 데이터

In [26]:
# 부호가 모든 같은 경우가 있는가? 요모두 동점인 경우가 생길것이다.
for i in range(len(X_test)):
    # ~. decision_function을 이용하면 해당 데이터가 하이퍼플레인으로부터 얼마나 떨어져있는 지 거리가 나온다 다음은 그 값의 부호를 이용해
    # 모두가 동점인 경우가 있는지 출력하는 함수 
    if ((np.sign(svm_1.decision_function(X_test)[i]) == np.sign(svm_2.decision_function(X_test)[i]))
    and (np.sign(svm_2.decision_function(X_test)[i]) == np.sign(svm_3.decision_function(X_test)[i]))):
        print(i)

3
17
18


4번째, 18번째, 19번째 데이터가 모두 동점인 경우다. 위에서 0이라고 표시된 인덱스와 같음

In [30]:
# 동점인 경우는 결정함수값을 모두 출력해서 직접 비교하여 예측하기
for i in [3,17,18]:
    print(svm_1.decision_function(X_test)[i], svm_2.decision_function(X_test)[i], svm_3.decision_function(X_test)[i])

-0.5024882107322184 -0.36987092513822756 -0.13070947139128608
-0.562404550285589 -0.18705039695337913 -0.2533393758266559
-0.7271989224328648 -0.24458235766281572 -0.025197903756274675


가장 거리가 가까운(절댓값으로 비교) 값을 찾으면

인덱스 3: -0.13070947139128608(virginica)

인덱스 17: -0.18705039695337913(versicolor)

인덱스 18: -0.025197903756274675(virginica)

In [31]:
#직접 리스트에 넣어주기
group[3] = 'virginica'
group[17] = 'versicolor'
group[18] = 'virginica'

In [32]:
# 결정함수로 예측된 데이터들
group

['versicolor',
 'versicolor',
 'versicolor',
 'virginica',
 'virginica',
 'virginica',
 'setosa',
 'virginica',
 'setosa',
 'versicolor',
 'virginica',
 'setosa',
 'setosa',
 'virginica',
 'versicolor',
 'versicolor',
 'setosa',
 'versicolor',
 'virginica',
 'virginica',
 'setosa',
 'virginica',
 'versicolor',
 'versicolor',
 'virginica',
 'setosa',
 'setosa',
 'virginica',
 'virginica',
 'versicolor']

In [37]:
# 원래 라이브러리가 제공하는 멀티클래스 SVM 결과
from sklearn.model_selection import train_test_split #데이터셋 분리
from sklearn import metrics
X_train_2, X_test_2, y_train_2, y_test_2 = train_test_split(X, y, test_size=0.2, random_state=48)

svm_4 = SVC(kernel ='rbf', C = 5, gamma = 5)
svm_4.fit(X_train_2, y_train_2)
y_pred = svm_4.predict(X_test_2)

metrics.accuracy_score(y_test_2,y_pred)

0.8333333333333334

In [36]:
# 직접 구현한 함수로 예측한 결과
metrics.accuracy_score(y_test,group)

0.8666666666666667

라이브러리 결과보다 0.03정도 더 정확도가 높다